Style guide: _tac instantiations and attributes

Xaphiosis opened the pull request:
Adds two sections:
- Fact transformers and attributes (due to some experience with PRs this year)
- Wrapping/indenting complex rule instantiations, usually as part of `_tac` methods
  - as per https://github.com/seL4/l4v/issues/746

I also added consistent double-newline before sections, and added some more emphasis on 2 being the standard indent at the start of the indentation section.

Commits are independent so can be reviewed separately. Let the bikeshedding begin.

lsf37 reviewed (approved):
Happy with this colour of the bike shed as it is. Thanks for adding these.

michaelmcinerney commented on docs/Style.thy:
```diff
@@ -87,11 +88,17 @@ text \<open>
+  2 spaces when indenting something with respect to its container (see ``General layout`` in
+  ``Other`` section).
+
+  In apply style proofs we indent by 2 spaces, and add an additional space for every additional
```
I know this is how it was phrased in the previous version, but would it be a bit more clear to say after what's here something like "for example, a command which applies when there are 3 subgoals should be indented by 4 spaces".

Xaphiosis replied:
Sure, I can add that while we're here.

Xaphiosis replied:
Added, let me know if it's clear now.

michaelmcinerney replied:
Yep, thanks.

michaelmcinerney reviewed (approved):
Looks good

corlewis commented on docs/Style.thy:
```diff
@@ -87,11 +88,18 @@ text \<open>
-  Isabelle code is much easier to maintain when indented consistently. In apply style proofs we
-  indent by 2 spaces, and add an additional space for every additional subgoal.
+  Isabelle code is much easier to maintain when indented consistently.
+  When in doubt, and not constrained by vertically aligning items or subgoal-count offsets, use
```
Extremely minor grammar nitpick
```suggestion
  When in doubt and not constrained by vertically aligning items or subgoal-count offsets, use
```

Xaphiosis replied:
*shrug* ... if it helps, I'll add the comma

corlewis replied:
I'm suggesting to remove the comma, I don't think it's right.

Xaphiosis replied:
Oh, yes, will do then

corlewis commented on docs/Style.thy:
```diff
@@ -359,6 +359,66 @@ term "
+section \<open>Wrapping/indenting complex rule instantiations, usually as part of @{text "_tac"} methods\<close>
+
+text \<open>
+  For simple instantiations, they can fit it on one line, like so.
```
This sentence is a bit unclear to me. Are you saying that simple instantiations that can fit on one line should look like the following?

Xaphiosis replied:
Yes. I should make that more explicit.

corlewis commented on docs/Style.thy:
```diff
@@ -359,6 +359,66 @@ term "
+
+lemma
+  "\<lbrakk>a; b\<rbrakk> \<Longrightarrow> a \<and> b"
+  apply (frule_tac P=a and Q="id b" in conjI) \<comment> \<open>when @{text a} and @{text b} contain bound vars\<close>
```
It's not clear to me what the intention of the comment is here. Is it about using the `_tac` form instead of `where`?

Xaphiosis replied:
Yes, that is correct. I want to avoid guiding people towards writing out the rule_tac version when a `fact[where ...]` will suffice. The rule_tac version is less generic, as it applies only to the `_tac`s, but the attribute version has many more uses.

corlewis replied:
I agree with that goal but then I think that the comment needs to be more explicit. As it is the mention of bound variables seems to come out of nowhere and the alternative attribute version isn't mentioned.

However, I'm then struggling to think of a way to keep it short. Would it make sense to add it as a note instead, with something vaguely like
`@{text "_tac"} methods as shown here are generally used when the instantiations contain bound vars, otherwise instantiations with the @{text "[where ...]" attribute are preferred as they are more general.`

Xaphiosis replied:
Ok, tuned, how does the latest commit look to you?

corlewis replied:
One potential typo, otherwise looks good to me!

corlewis commented on docs/Style.thy:
```diff
@@ -226,6 +226,43 @@ lemma test_lemma3:
+                                simplified],
+                 simplified] \<comment> \<open>extreme wrapping case\<close>
+
+text \<open>
+  There is an important principle here: telling apart transformed/attributed facts from unaltered
+  facts at a glance. In other words avoid:\<close>
```
This is a very useful comment, good thinking to include it. People are generally much better at following rules if they know the intention behind it, as opposed to it just being some arbitrary thing that they need to remember.

corlewis reviewed (approved):
Some very minor bikeshedding, thanks for updating the style guide for these!

Xaphiosis commented:
I fixed up non-proof items (including reverting the copyright changes) and added an explanation for the locale hierarchy based partly on Gerwin's explanation. Now I'll be looking at the proof changes, which will take longer.

corlewis commented on docs/Style.thy:
```diff
@@ -351,6 +396,72 @@ term "
+  (@{text "fact[where ...]"}) as they are more general.\<close>
+
+text \<open>
+  For simple instantiations which can fit it on one line, style them as follows.
```
```suggestion
  For simple instantiations which can fit on one line, style them as follows.
```
